lib/virtual-fs/src/lib.rs

```diff
@@ -0,0 +1,319 @@
+    /// The connection was reset
+    #[error("connection reset")]
+    ConnectionReset,
+    /// The operation was interrupted before it could finish
+    #[error("operation interrupted")]
+    Interrupted,
+    /// Invalid internal data, if the argument data is invalid, use `InvalidInput`
+    #[error("invalid internal data")]
+    InvalidData,
+    /// The provided data is invalid
+    #[error("invalid input")]
+    InvalidInput,
+    /// Could not perform the operation because there was not an open connection
+    #[error("connection is not open")]
+    NotConnected,
+    /// The requested file or directory could not be found
+    #[error("entity not found")]
+    EntityNotFound,
+    /// The requested device couldn't be accessed
+    #[error("can't access device")]
+    NoDevice,
+    /// Caller was not allowed to perform this operation
+    #[error("permission denied")]
+    PermissionDenied,
+    /// The operation did not complete within the given amount of time
+    #[error("time out")]
+    TimedOut,
+    /// Found EOF when EOF was not expected
+    #[error("unexpected eof")]
+    UnexpectedEof,
+    /// Operation would block, this error lets the caller know that they can try again
+    #[error("blocking operation. try again")]
+    WouldBlock,
+    /// A call to write returned 0
+    #[error("write returned 0")]
+    WriteZero,
+    /// Some other unhandled error. If you see this, it's probably a bug.
+    #[error("unknown error found")]
+    UnknownError,
+}
+
+impl From<io::Error> for FsError {
+    fn from(io_error: io::Error) -> Self {
+        match io_error.kind() {
+            io::ErrorKind::AddrInUse => FsError::AddressInUse,
+            io::ErrorKind::AddrNotAvailable => FsError::AddressNotAvailable,
+            io::ErrorKind::AlreadyExists => FsError::AlreadyExists,
+            io::ErrorKind::BrokenPipe => FsError::BrokenPipe,
+            io::ErrorKind::ConnectionAborted => FsError::ConnectionAborted,
+            io::ErrorKind::ConnectionRefused => FsError::ConnectionRefused,
+            io::ErrorKind::ConnectionReset => FsError::ConnectionReset,
+            io::ErrorKind::Interrupted => FsError::Interrupted,
+            io::ErrorKind::InvalidData => FsError::InvalidData,
+            io::ErrorKind::InvalidInput => FsError::InvalidInput,
+            io::ErrorKind::NotConnected => FsError::NotConnected,
+            io::ErrorKind::NotFound => FsError::EntityNotFound,
+            io::ErrorKind::PermissionDenied => FsError::PermissionDenied,
+            io::ErrorKind::TimedOut => FsError::TimedOut,
+            io::ErrorKind::UnexpectedEof => FsError::UnexpectedEof,
+            io::ErrorKind::WouldBlock => FsError::WouldBlock,
+            io::ErrorKind::WriteZero => FsError::WriteZero,
+            io::ErrorKind::Other => FsError::IOError,
+            // if the following triggers, a new error type was added to this non-exhaustive enum
+            _ => FsError::UnknownError,
+        }
+    }
+}
```

Comment:
Can we have a `MemoryFS` also with tests?

Reply:
Yep, working on  it